File: dev/test_data/cpp/test_main/aas_core_meta.v3/input/snippets/verification/ID_shorts_of_variables_are_unique.cpp
bool IdShortsOfVariablesAreUnique(
  const common::optional<
    std::vector<
      std::shared_ptr<types::IOperationVariable>
    >
  >& input_variables,
  const common::optional<
    std::vector<
      std::shared_ptr<types::IOperationVariable>
    >
  >& output_variables,
  const common::optional<
    std::vector<
      std::shared_ptr<types::IOperationVariable>
    >
  >& inoutput_variables
) {
  std::set<std::wstring> id_short_set;
  
  if (input_variables.has_value()) {
    const std::vector<
      std::shared_ptr<types::IOperationVariable>
    >& variables = *input_variables;
    
    for (const std::shared_ptr<types::IOperationVariable>& variable : variables) {
      const common::optional<std::wstring>& id_short(
        variable->value()->id_short()
      );
      
      if (id_short.has_value()) {
        if (id_short_set.find(*id_short) != id_short_set.end()) {
          return false;
        }
        
        id_short_set.insert(*id_short);
      }
    }
  }
  
  if (output_variables.has_value()) {
    const std::vector<
      std::shared_ptr<types::IOperationVariable>
    >& variables = *output_variables;
    
    for (const std::shared_ptr<types::IOperationVariable>& variable : variables) {
      const common::optional<std::wstring>& id_short(
        variable->value()->id_short()
      );
      
      if (id_short.has_value()) {
        if (id_short_set.find(*id_short) != id_short_set.end()) {
          return false;
        }
        
        id_short_set.insert(*id_short);
      }
    }
  }
  
  if (inoutput_variables.has_value()) {
    const std::vector<
      std::shared_ptr<types::IOperationVariable>
    >& variables = *inoutput_variables;
    
    for (const std::shared_ptr<types::IOperationVariable>& variable : variables) {
      const common::optional<std::wstring>& id_short(
        variable->value()->id_short()
      );
      
      if (id_short.has_value()) {
        if (id_short_set.find(*id_short) != id_short_set.end()) {
          return false;
        }
        
        id_short_set.insert(*id_short);
      }
    }
  }

  return true;
}

File: dev/test_data/main/python/expected/aas_core_meta.v3/input/snippets/Verification/ID_shorts_of_variables_are_unique.py
def id_shorts_of_variables_are_unique(
    input_variables: Optional[List[aas_types.OperationVariable]],
    output_variables: Optional[List[aas_types.OperationVariable]],
    inoutput_variables: Optional[List[aas_types.OperationVariable]],
) -> bool:
    """
    Check that the :py:attr:`.types.Referable.id_short`'s among all the
    :paramref:`input_variables`, :paramref:`output_variables`
    and :paramref:`inoutput_variables` are unique.
    """
    id_short_set = set()
    if input_variables is not None:
        for variable in input_variables:
            if variable.value.id_short is not None:
                if variable.value.id_short in id_short_set:
                    return False

                id_short_set.add(variable.value.id_short)
    if output_variables is not None:
        for variable in output_variables:
            if variable.value.id_short is not None:
                if variable.value.id_short in id_short_set:
                    return False

                id_short_set.add(variable.value.id_short)
    if inoutput_variables is not None:
        for variable in inoutput_variables:
            if variable.value.id_short is not None:
                if variable.value.id_short in id_short_set:
                    return False

                id_short_set.add(variable.value.id_short)
    return True
